### DISCUSSION TITLE
Export metrics as prometheus format

### DISCUSSION THREAD

**Discussion by elonzh** [meilisearch/product#146] [Feedback & Feature Proposal]
**Is your feature request related to a problem? Please describe.**

MeiliSearch provides stats API but can't integrate with monitor system directly like prometheus.

**Describe the solution you'd like**

Provide a Prometheus metrics API out of box.

**Describe alternatives you've considered**

Write an [exporter](https://prometheus.io/docs/instrumenting/exporters/)

**Additional context**

- https://github.com/tikv/rust-prometheus
- https://grafana.com/

## Addition by @gmourier

#91

**Comment by curquiza** [comment]
@gmourier 😇

**Comment by gmourier** [comment]
Hello @elonzh! Thanks for the feedback, can you tell me more about the metrics you would like to monitor and why? `/stats` is quite simple at the moment.

We've added a product card about [prometheus](https://roadmap.meilisearch.com/c/102-prometheus-metrics) on our interactive roadmap so you can vote for it!

**Comment by elonzh** [comment]
query/update performance metrics are important for production use.

**Comment by gmourier** [comment]
Thank you, I have an additional question to better understand. Is the goal to have a general view of the good shape condition of the search engine (Alerting purpose) or to perform analysis on `search` query/results to improve the behavior of the search engine to meet business criteria? I have the impression that this is the first one but maybe other needs can be expressed!

**Comment by elonzh** [comment]
Yes, you are right!

**Comment by gmourier** [comment]
Hello @elonzh

I see that you considered writing an exporter. Could I know why you didn't consider this solution in the end? We started this subject, and we are wondering whether to write an exporter or provide an endpoint directly returning data formatted for Prometheus.

I would like to know if you see advantages/disadvantages between the two solutions.

Thanks!

**Reply by elonzh** [reply]
Typically providing an endpoint directly is a better solution for modern projects(especially for projects that already have http endpoints), it's easy to use, and does not need data conversion, but it's invasive and the project maintainer has the responsibility to implement it.

**Reply by elonzh** [reply]
Also, if the goal of your project is to support interfacing with various monitoring systems, it may be better to implement an easy-to-convert data interface. For example, in Java, we can use [micrometer](https://micrometer.io/docs) to adapt metrics to various monitoring systems.

**Reply by elonzh** [reply]
You can check out [exporters](https://prometheus.io/docs/instrumenting/exporters/#third-party-exporters) listed in Prometheus docs, you will find most exporter developers are not the project maintainer and some projects indeed need an exporter because they are not HTTP services or they already have a metric system.

**Reply by gmourier** [reply]
Thanks, @elonzh this is super useful.

**Comment by gmourier** [comment]
Hey everyone, I'm locking that conversation in favor of #625. 

That discussion is made to centralize every feedback around the `/metrics` Prometheus endpoint Meilisearch will expose as an experimental feature with `v1.1`
